### PR TITLE
feat: explicit inflight references [WIP]

### LINKED ISSUE #1595: Cannot use `Resource` APIs in Wing
### I tried this:

For example, I want to be able to update the [`display`](https://github.com/winglang/wing/blob/5332d6ecdb16362ced14b709df916f16e0cb36cd/libs/wingsdk/src/core/resource.ts#L30) properties of the resource:

```js
let b = new cloud.Bucket();
b.display.title = "MyBucket";
```

### I expected this:

To work

### Instead, this happened:

Unknown symbol.

### Component:

Compiler

### Wing Version:

0.5.37

### Node.js Version:

_No response_

### Platform(s):

_No response_

### Anything else?

@yoav-steinberg said:

> The JSII importer respects base class methods, but `core.Resource` is a special case. It's actually never imported.

Related: https://github.com/winglang/wing/issues/261

### Community Note

---
* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### PR DESCRIPTION
**Pull request by eladb**
Require that all references between inflight methods and preflight resources will be recorded explicitly via a call to a new API at the `Resource` level called `inflights.add()`.

For example:

```js
bring cloud;

resource MyResource {
  b: cloud.Bucket;
  init() { 
    this.b = new cloud.Bucket(); 
  }

  inflight f1() { 
    this.b.put("hello", "world");
  }
}

let r = new MyResource();
new cloud.Function(inflight () => { r.f1(); });
```

If the above is compiled, the following error will be emitted:

```
preflight error: root/MyResource: unable to determine reference between "f1" and "this.b". Use `this.inflight.add("f1", ref: "this.b", op: "<OP>")`.
```

To resolve this, users will need to explicitly add the following to the resource initializer:

```js
init() {
  this.b = new cloud.Bucket();
  this.inflights.add("f1", ref: "this.b", op: "put");
}
```

This records the reference between `f1` and `this.b` and also indicates that the operation performed is `put()`.

### Implementation Notes

1. Currently, we don't really analyze inflight methods for captures - we simply consider the cartesian product of all preflight fields with all inflight methods. This is not realistic.
2. The `core.Resource` class is now imported from the SDK and user-defined `resource`s now implicitly extend it. This fixes #1595. Since we currently don't support importing interfaces, I had to tweak the API a little and get rid of the `IInflightHost` interface (just use `Resource` instead). This is reasonable at this point and not a huge loss at this point. We can restore this later.
3. The error above is emitted *only* when the resource is being referenced by a `cloud.Function`, which is not ideal, but maybe that's good enough for now. Not sure...

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

**Comment by ShaiBer** [comment]
Is this a temporary measure for now, or do you plan that users will need to use this api for such simple cases in the Beta as well? I know you were talking about it today but don't remember what was said exactly.
I think we should do whatever we can to avoid this in Beta because it takes a lot of the magic away from wing and we cannot really say the compiler takes care of IAM policies if developers need to spoon feed it. In addition, it feels like a very bad DX to me.
I think that even the broad permissions we have today are better because we can say that the compiler at least does the work automatically and we just need to fine tune it to be more narrow. Whereas if users need the new api it feels like we are further away from a compiler that can generate the IAM policies automatically and reduce the complexity of the cloud for us.
But maybe we can eat our cake and leave it whole too: why not have the compiler emit a warning in the above example instead of an error which says that unless we use the API to narrow the permissions it will generate broad permissions for us? Then, as the compiler advances, it will learn to give least privilege permissions on such trivial cases and keep the warning for the more complex cases where it still cannot resolve the permissions.

**Review by Chriscbr — COMMENTED** [review]
SDK code looks good to me 👍

I realize this could be a bit of a hit to the user experience, but I'm personally OK with the tradeoffs (or at least giving this DX a try). We can always make a new PR that turns back on the automatic cartesian-product-default-annotations -- or even put it behind a compiler flag of some sort.

Additional question - do you think it's feasible to make the API look like this?
```js
this.inflights.add("put1", { ref: this.b1, op: "put" });
```
(directly passing `this.b1` instead of a string)? I remember there was some kind of technical reason we needed strings originally but I'm not sure if it still applies now...

**Comment by ekeren** [comment]
> if the above is compiled, the following error will be emitted:
>> preflight error: root/MyResource: unable to determine reference between "f1" and "this.b". Use `this.inflight.add("f1", ref: "this.b", op: "<OP>")`.

I suggest we do not enforce this restriction at the moment, it break our code examples (including our canonical one) and takes the edge of the current DX with a WTF moment.

If we want we can have an env flag to WINGC_STRICT_BIND=true or something that will **not be the default**

**Comment by ShaiBer** [comment]
> > if the above is compiled, the following error will be emitted:
> > > preflight error: root/MyResource: unable to determine reference between "f1" and "this.b". Use `this.inflight.add("f1", ref: "this.b", op: "<OP>")`.
> 
> I suggest we do not enforce this restriction at the moment, it break our code examples (including our canonical one) and takes the edge of the current DX with a WTF moment.
> 
> If we want we can have an env flag to WINGC_STRICT_BIND=true or something that will **not be the default**

Makes sense. Even if they don't use the flag, developers will be able to use the API in cases they want to add more restrictive permissions

File: tools/hangar/src/__snapshots__/invalid.test.ts.snap [hunk 1]
```diff
@@ -197,19 +196,19 @@ Error at ../../../../examples/tests/invalid/unimplemented_grammar.w:5:9 | expres
...
 "Compilation failed with 2 error(s)
-Error at ../../../../examples/tests/invalid/use_before_defined.w:1:10 | Unknown symbol \\"y\\"
+Error at ../../../../examples/tests/invalid/use_before_defined.w:1:10 | Unknown symbol \\"y\\". Did you mean? \\"assert\\",\\"panic\\",\\"print\\",\\"std\\",\\"throw\\"
 Error at ../../../../examples/tests/invalid/use_before_defined.w:5:12 | Symbol \\"x\\" used before being defined"
 `;
 
 exports[`variable_scoping.w > wing test (--target sim) - invalid > stderr 1`] = `
 "Compilation failed with 1 error(s)
-Error at ../../../../examples/tests/invalid/variable_scoping.w:6:11 | Unknown symbol \\"x\\""
+Error at ../../../../examples/tests/invalid/variable_scoping.w:6:11 | Unknown symbol \\"x\\". Did you mean? "
```
[the remark this answers is not in the document]
Comment (eladb) [reply tools/hangar/src/__snapshots__/invalid.test.ts.snap:211]: Good call

File: libs/wingc/src/type_check.rs [hunk 2]
```diff
@@ -1605,6 +1607,20 @@ impl<'a> TypeChecker<'a> {
...
+					let resource_fqn = &format!("{}.{}", WINGSDK_ASSEMBLY_NAME, WINGSDK_RESOURCE);
+					let resource_base_type = self.types.libraries
+						.lookup_nested_str(resource_fqn, true, None)
+						.expect("core.Resource not found")
+						.as_type()
+						.expect("core.Resource is not a type");
+
+					let resource_base_class = resource_base_type
+						.as_resource()
+						.expect("core.Resource is not a resource");
```
[the remark this answers is not in the document]
Comment (eladb) [reply libs/wingc/src/type_check.rs:1621]: I learned from the best (@yoav-steinberg)(

**Comment by eladb** [comment]
@ShaiBer i don't intend to merge this before we have basic capture analysis. I synced today with @yoav-steinberg and he is on it.

This means that the compiler will basically identify most references and these errors will only be left for the cases where the compiler cannot identify the exact operation being performed.

Stay tuned.

**Comment by ShaiBer** [comment]
> @ShaiBer i don't intend to merge this before we have basic capture analysis. I synced today with @yoav-steinberg and he is on it.
> 
> This means that the compiler will basically identify most references and these errors will only be left for the cases where the compiler cannot identify the exact operation being performed.
> 
> Stay tuned.

Sounds great @eladb !
Still maybe broad auto permissions + a warning will be a better DX, but we have time to discuss :)

**Pull request closed without merging**